sycl/test/devicelib/cmath_test.cpp

```diff
@@ -0,0 +1,85 @@
+// UNSUPPORTED: windows
+// RUN: %clangxx -fsycl -c %s -o %t.o
+// RUN: %clangxx -fsycl %t.o %llvm_build_libs_dir/libsycl-cmath.o -o %t.out
```

Comment:
As this is standard library functionality, it should be supported by the compiler, the same way it's supported for regular C/C++. Do you agree?
If so, device side linking must be done implicitly.
+@mdtoguchi.

Reply:
Yes, I have discussed with Andrew. We both think this should be done implicitly just link libc/libm.a in normal CPU program.

Reply:
Are you going to fix that?
I don't fully understand how many device libraries you are planning to have, but ideally we should avoid changing driver each time new device library is added.

Reply:
Hi, Alexey. For comlex/cmath, we don't need any other device libraries. However, if any other developers want to enable other functionalities they may want to add other device libraries.

Reply:
As I mentioned here: https://github.com/intel/llvm/pull/975#discussion_r364734220, we should document our design decision. 
It's not clear to me, why we keep double precision math functions as part of math library. 
NOTE: not all accelerators has native support for double precision arithmetic.
Does it make sense to move double precision functions to a separate library?

Reply:
As to original question with driver changes - they can be done in a separate patch.
Please, make sure this request is not lost if you decide to fix it later.

Reply:
Hi, Alexey.
To the driver changes, I think both libsycl-glibc.o and libsycl-complex.o/libsycl-math.o needs to be added and we need to make request to driver owner.
To double precision, I didn't get your point very well. Do you mean only float type math and complex functions are needed in fallback device libraries as some accelerators may not support double? If so, can we move all double type math/complex functions' implementation to libsycl-cmath.o and libsycl-complex.o? Then, those double type implementations will be linked with users' device code in compiling time and will exist in final device code only users really depend on them.
Thank you.

Reply:
> Do you mean only float type math and complex functions are needed in fallback device libraries as some accelerators may not support double?

No. I'm asking about the reasoning of splitting device library into different parts. It's not clear to me. I might imagine that it might be split differently. I can suggest anything only after I understand the logic behind the current proposal.

Reply:
Do you ask about why split the device libraries based on functionality(one for cmath and another for complex math) but not based on other criterion?

Reply:
Yes.
Also where I can find definition of "split ... based on functionality"? C/C++ specification?

Reply:
Hi, Alexey.
Before doing the cmath/complex support, I consulted with Andy and Andrew about the granularity of the device libraries and got the feedback that we can split all those functionalities into 2 device libraries: libsycl-cmath and libsycl-complex. And in fact, I think that the granularity for device libraries is a kind of trade off. In fact, for the most flexibility, we should provide one device library for each function provided, so backend compiler can choose any subset of all those functions to override. However, it is not practical to do so as we will have too many device libraries. Putting everything in just one device library is not a good idea too, as I mentioned before, backend compiler will lose other fallback device functions if only wants to support a subset of all functions. So, we have to do some trade off and I think "split...based on functionality" is not defined by any spec but a kind of experience. In CPU platform, we have many libraries such as libc, libm, libstdc++...., libc includes many functions used for C runtime, all math related functions are in libm and libstdc++ will include some functions  for std, we can use the similar way for device library. 
@andykaylor and @asavonic , do you have any comments?
Thank you very much.

Reply:
Right, the "split" is arbitrary, but we try to split how it makes sense. A separate library for each function does not make sense. A single library for *all* functions defeats the purpose of having extensions. C and C++ standard library groups functions and classes by purpose (math.h and compex.h), so we use this as a baseline.

Reply:
> C and C++ standard library groups functions and classes by purpose (math.h and compex.h), so we use this as a baseline.

I think something like this must be put into our architecture documentation, but with reference to the spec, rather than particular implementation. The spec defines headers, which slit stadard library declarations.

Reply:
> > C and C++ standard library groups functions and classes by purpose (math.h and compex.h), so we use this as a baseline.
> 
> I think something like this must be put into our architecture documentation, but with reference to the spec, rather than particular implementation. The spec defines headers, which slit stadard library declarations.

No strict rule is intended here. It is not clear why you want to define a formal criteria on how the split should be made.

Reply:
> > > C and C++ standard library groups functions and classes by purpose (math.h and compex.h), so we use this as a baseline.
> > 
> > 
> > I think something like this must be put into our architecture documentation, but with reference to the spec, rather than particular implementation. The spec defines headers, which slit stadard library declarations.
> 
> No strict rule is intended here. It is not clear why you want to define a formal criteria on how the split should be made.

I do not say it's "formal criteria", I want us to document this design decision to keep the implementation sane and consistent. This decision is not obvious from current implementation and if someone else will add new functionality the design must be clear.

Reply:
I think we need a device library, implemented as wrappers, for all of the functions that correspond to OpenCL built-ins. The fact that the devices may have native support for these is irrelevant. We need the library to perform the mapping from libm call to OpenCL built-in. The alternative is to perform this mapping in the compiler, but we've discussed that and decided the device library is preferable.

I think it would make sense to have a division of the device library for non-complex functions between those with OpenCL equivalents and those without. My suggestion for a separate complex library was based on the idea that some devices might make the decision that complex support was not a priority and the fallback implementations were acceptable.

Reply:
Already added the design decision in C-CXX-StandardLibrary.rst.